tezos-encoding/src/nom.rs

```diff
@@ -572,6 +573,56 @@ where
     }
 }
 
+#[derive(Debug, PartialEq)]
+/// Lazy, memoized deserialization of a dynamic block.
+pub struct LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    // Invariant: when the `deserialized` field is some, it's the
+    // result of deserialization of the `serialized` field using
+    // dynamic(T::nom_read)
+    serialized: &'a [u8],
+    deserialized: Option<T>,
+}
+
+impl<'a, T> NomReader<'a> for LazyCell<'a, T>
+where
+    T: NomReader<'a>,
+{
+    fn nom_read(input: &'a [u8]) -> NomResult<'a, Self> {
+        map(dynamic(rest), |serialized| LazyCell {
```

Comment:
does this need to require `dynamic` internally? It should delegate it perhaps to the underlying type?

(e.g. this will prevent using `LazyCell` with e.g. the  Rollup Inbox)

Reply:
IIUC, you suggest to assume that `T` itself uses a `dynamic` encoding and to keep the size header in the `serialized` field. I don't know if we can enforce statically that `T` is a dynamic encoding.